packages/services/src/spot-data-processor.service.ts

```diff
@@ -0,0 +1,1017 @@
+      const mapKey = `${address.toLowerCase()}:${chain}`;
+      const price = bulkPriceMap.get(mapKey);
+
+      if (price) {
+        priceMap.set(mapKey, price);
+        successCount++;
+      } else {
+        failCount++;
+      }
+    }
+
+    this.logger.debug(
+      `[SpotDataProcessor] Bulk price fetch complete: ${successCount}/${tokens.length} tokens priced` +
+        (failCount > 0 ? `, ${failCount} failed` : ""),
+    );
+
+    return priceMap;
+  }
+
+  /**
+   * Process data for a single agent
+   *
+   * Gap Prevention Strategy:
+   * This method uses incremental block-based syncing with overlapping scans to guarantee
+   * no missed transactions even when partial failures occur.
+   *
+   * How it works:
+   * 1. Query latest processed block number per chain from trades table
+   * 2. Sync from latestBlock (with overlap), NOT latestBlock+1
+   * 3. Unique constraint (txHash, competitionId, agentId) prevents duplicates
+   * 4. Duplicate check BEFORE balance updates (efficient)
+   *
+   * Why overlap prevents gaps:
+   * - Block 1000: [Trade A ✅, Trade B ❌, Trade C ✅] - B fails, A & C succeed
+   * - MAX(block_number) = 1000 (from A or C)
+   * - Next sync from block 1000 (overlap) → Retries B ✅
+   * - Duplicates from A & C are caught by duplicate check (no balance impact)
+   *
+   * Cost vs Safety:
+   * - Overhead: ~1 block worth of duplicate checks per sync (minimal)
+   * - Benefit: Zero transaction gaps, automatic retry of failures
+   *
+   * @param agentId Agent ID
+   * @param competitionId Competition ID
+   * @param walletAddress Agent's wallet address
+   * @param provider The spot live data provider
+   * @param allowedTokens Map of chain to Set of allowed token addresses
+   * @param tokenWhitelistEnabled Whether token filtering is enabled
+   * @param chains Array of chains to scan
+   * @param competitionStartDate Competition start date (fallback for first sync)
+   * @returns Processing result
+   */
+  async processAgentData(
+    agentId: string,
+    competitionId: string,
+    walletAddress: string,
+    provider: ISpotLiveDataProvider,
+    allowedTokens: Map<string, Set<string>>,
+    tokenWhitelistEnabled: boolean,
+    chains: SpecificChain[],
+    competitionStartDate: Date,
+  ): Promise<AgentSpotSyncResult> {
+    if (!provider) {
+      throw new Error("[SpotDataProcessor] Provider is required");
+    }
+
+    const startTime = Date.now();
+
+    try {
+      this.logger.info(
+        `[SpotDataProcessor] Processing agent ${agentId} for competition ${competitionId}`,
+      );
+
+      // 1. Determine sync start point per chain for incremental processing
+      // Uses sync state + bounded overlap for gap-free syncing with retry capability
+      const allTrades: OnChainTrade[] = [];
+      const syncStateUpdates: Array<{
+        chain: SpecificChain;
+        highestBlock: number;
+      }> = [];
+
+      const RETRY_WINDOW_BLOCKS = 10; // Retry last 10 blocks for transient failures
```

Comment:
Why 10 here?

Reply:
10 blocks provides a time-based retry window that varies by chain:
- Base: ~20 seconds (2s/block × 10)
- Ethereum: ~120 seconds (12s/block × 10)

This catches transient price API failures (rate limits, network blips) which typically resolve within 1-2 minutes, while keeping overhead bounded. Trade-offs:
- <10: Insufficient recovery time for transient failures
- >10: Excessive duplicate checks, wasted API calls
- 10: Balances retry capability with efficiency